velox/functions/prestosql/DateTimeImpl.h

```diff
@@ -187,4 +184,120 @@ FOLLY_ALWAYS_INLINE Timestamp addToTimestamp(
+      case DateTimeUnit::kMillisecond: {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(diffMilli)
+            .count();
+      }
+      case DateTimeUnit::kSecond: {
+        return std::chrono::duration_cast<std::chrono::seconds>(diffMilli)
+            .count();
+      }
+      case DateTimeUnit::kMinute: {
+        return std::chrono::duration_cast<std::chrono::minutes>(diffMilli)
+            .count();
+      }
+      case DateTimeUnit::kHour: {
+        return std::chrono::duration_cast<std::chrono::hours>(diffMilli)
+            .count();
+      }
+      case DateTimeUnit::kDay: {
+        return std::chrono::duration_cast<date::days>(diffMilli).count();
+      }
+      default:
+        VELOX_UNREACHABLE();
+    }
+  }
+
+  // Month, quarter and year are not uniformly incremented.
+  const date::year_month_day fromCalDate(
+      std::chrono::floor<date::days>(fromTimepoint));
+  const date::year_month_day toCalDate(
+      std::chrono::floor<date::days>(toTimepoint));
+  const uint64_t fromTimeInstantOfDay =
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          fromTimepoint - std::chrono::floor<date::days>(fromTimepoint))
+          .count();
+  const uint64_t toTimeInstantOfDay =
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          toTimepoint - std::chrono::floor<date::days>(toTimepoint))
+          .count();
+  const uint32_t fromDay = static_cast<unsigned>(fromCalDate.day()),
+                 fromMonth = static_cast<unsigned>(fromCalDate.month());
+  const uint32_t toDay = static_cast<unsigned>(toCalDate.day()),
+                 toMonth = static_cast<unsigned>(toCalDate.month());
+
+  if (unit == DateTimeUnit::kMonth || unit == DateTimeUnit::kQuarter) {
+    int64_t diff = (int(toCalDate.year()) - int(fromCalDate.year())) * 12 +
+        int(toMonth) - int(fromMonth);
+
+    const date::year_month_day toCalLastDay(
+        toCalDate.year() / toCalDate.month() / date::last);
+    if ((toDay != static_cast<unsigned>(toCalLastDay.day()) &&
+         fromDay > toDay) ||
+        (fromDay == toDay && fromTimeInstantOfDay > toTimeInstantOfDay)) {
+      diff--;
+    }
+    return unit == DateTimeUnit::kMonth ? diff : diff / 3;
+  }
+
+  if (unit == DateTimeUnit::kYear) {
+    int64_t diff = (toCalDate.year() - fromCalDate.year()).count();
+
+    if (fromMonth > toMonth ||
+        (fromMonth == toMonth && fromDay > toDay && toMonth != 2 &&
+         toDay != 28) ||
+        (fromMonth == toMonth && fromDay == toDay &&
+         fromTimeInstantOfDay > toTimeInstantOfDay)) {
+      diff--;
+    }
+    return diff;
+  }
+
+  VELOX_UNREACHABLE();
+}
+
+FOLLY_ALWAYS_INLINE
+int64_t
+diffDate(const DateTimeUnit unit, const Date& fromDate, const Date& toDate) {
+  if (fromDate == toDate) {
+    return 0;
+  }
+
+  if (toDate < fromDate) {
+    return -diffDate(unit, toDate, fromDate);
+  }
+
+  return diffTimestamp(
+      unit,
+      Timestamp(fromDate.days() * 86'400, 0),
```

Comment:
Abstract this as a constant somewhere. Even using Date.h might be okay as we use this number in many places.

Reply:
Moving those time-related constants that include kSecsPerDay from TimestampConversion.cpp to TimestampConversion.h, and include TimestampConversion.h here.